## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [21]:
import pandas as pd

file_path = '/Users/desmonddavis/Desktop/pandas-challenge-1/Resources/client_dataset.csv'
data = pd.read_csv('Resources/client_dataset.csv')


In [22]:
# View the column names in the data

print("column names:")
print(data.columns)

column names:
Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [23]:
# Use the describe function to gather some basic statistics
print("\nData Description:")
print(data.describe())



Data Description:
          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.

In [24]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [25]:
# What three item categories had the most entries?
print("\nTop 3 Item Categories by Entries:")
category_counts = data['category'].value_counts()
print(category_counts.head(3))


Top 3 Item Categories by Entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [26]:
# For the category with the most entries, which subcategory had the most entries?
top_category = category_counts.idxmax()
subcategory_counts = data[data['category'] == top_category]['subcategory'].value_counts()
print(f"\nSubcategory with most entries in '{top_category}':")
print(subcategory_counts.idxmax())



Subcategory with most entries in 'consumables':
bathroom supplies


In [40]:
# Which five clients had the most entries in the data?
print("\nTop 5 Clients with Most Entries:")
client_counts = data['client_id'].value_counts()
top_5_clients = client_counts.head(5)
print(top_5_clients)


Top 5 Clients with Most Entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [41]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_clients.index.tolist()


In [42]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_5_client_ids[0]
total_units_top_client = data[data['client_id'] == top_client_id]['qty'].sum()
print(f"\nTotal units ordered by client {top_client_id}: {total_units_top_client}")


Total units ordered by client 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [47]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
data['subtotal'] = data['unit_price'] * data['qty']
print(data[['unit_price', 'qty', 'subtotal']].head())



   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [52]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
print("\nColumn Names in Dataset:")
print(data.columns)



Column Names in Dataset:
Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Subtotal', 'subtotal'],
      dtype='object')


In [57]:
# Update based on actual column name for weight
# Replace 'weight_pounds' with the actual column name in your dataset, e.g., 'weight'
data['shipping_price'] = data['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
print(data.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Subtotal', 'subtotal', 'shipping_price'],
      dtype='object')


In [63]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
data['total_price'] = (data['subtotal'] + data['shipping_price'])  * 1.0925
print(data.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Subtotal', 'subtotal', 'shipping_price', 'total_price'],
      dtype='object')


In [66]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
data['line_cost'] = (data['unit_cost'] * data['qty']) + data['shipping_price']
print(data.columns)



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Subtotal', 'subtotal', 'shipping_price', 'total_price',
       'line_cost'],
      dtype='object')


In [67]:
# Create a column for the profit of each line using line cost and line price
data['profit'] = data['total_price'] - data['line_cost']
print(data.head())



       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105   

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [72]:
# Check your work using the totals above
order_totals = data.groupby('order_id')['total_price'].sum()
print("\nConfirm Order Totals:")


Confirm Order Totals:


In [73]:

import pandas as pd
order_ids = ['2742071', '2173913', '6128929']
for order_id in order_ids:
    if order_id in order_totals.index.astype(str):
        print(f"Order ID {order_id}:", round(order_totals.loc[int(order_id)], 2))
    else:
        print(f"Order ID {order_id} not found.")

Order ID 2742071: 139844.51
Order ID 2173913: 119838.36
Order ID 6128929: 752748.75


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [76]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_client_data = data[data['client_id'].isin(top_5_client_ids)]
summary = top_5_client_data.groupby('client_id').agg(
    total_units=('qty', 'sum'),
    total_shipping=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
)
print("\nSummary for Top 5 Clients:")
print(summary)


Summary for Top 5 Clients:
           total_units  total_shipping  total_revenue  total_profit
client_id                                                          
24741           239862          9365.6   7.667848e+07  3.610666e+07
33615            64313         12609.4   6.392918e+06  2.033980e+06
38378            73667         11895.0   9.172866e+06  2.955706e+06
46820            75768         11094.8   8.006333e+06  2.589495e+06
66037            43018         10017.3   8.746255e+06  3.126907e+06


In [77]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
total_revenue_million = summary['total_revenue'] / 1e6
total_profit_million = summary['total_profit'] / 1e6
summary['total_revenue_millions'] = total_revenue_million
summary['total_profit_millions'] = total_profit_million
summary = summary[['total_units', 'total_shipping', 'total_revenue_millions', 'total_profit_millions']]
summary = summary.sort_values(by='total_profit_millions', ascending=False)

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [80]:
print("\nSummary for Top 5 Clients:")
print(summary)

print("\nFindings:")
print("The top 5 clients accounted for significant revenue and profit, indicating their importance in business performance. Further analysis could explore targeted strategies for these clients.")


Summary for Top 5 Clients:
           total_units  total_shipping  total_revenue_millions  \
client_id                                                        
24741           239862          9365.6               76.678479   
66037            43018         10017.3                8.746255   
38378            73667         11895.0                9.172866   
46820            75768         11094.8                8.006333   
33615            64313         12609.4                6.392918   

           total_profit_millions  
client_id                         
24741                  36.106662  
66037                   3.126907  
38378                   2.955706  
46820                   2.589495  
33615                   2.033980  

Findings:
The top 5 clients accounted for significant revenue and profit, indicating their importance in business performance. Further analysis could explore targeted strategies for these clients.
